src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -346,5 +353,64 @@ private BoundExpression RewriteStringConcatInExpressionLambda(CSharpSyntaxNode s
 
             return new BoundBinaryOperator(syntax, operatorKind, loweredLeft, loweredRight, default(ConstantValue), method, default(LookupResultKind), type);
         }
+
+        /// <summary>
+        /// Checks whether the expression represents a boxing conversion of a value type.
+        /// If it does, it returns a synthesized ToString call for that value type instance.
+        /// Otherwise, it returns the original expression.
+        /// </summary>
+        private BoundExpression ConvertExpressionToStringIfValueType(CSharpSyntaxNode syntax, BoundExpression expr)
+        {
+            // The lowered expression coming in will be a boxing conversion if the input is a value type.
+            // When passed to String.Concat, the resulting object will have ToString called on it.
+            // We can optimize away the boxing allocation by just calling ToString on the value type directly.
+
+            if (expr.Kind == BoundKind.Conversion)
+            {
+                BoundConversion conv = (BoundConversion)expr;
+                if (conv.ConversionKind == ConversionKind.Boxing)
+                {
+                    BoundExpression operand = conv.Operand;
+                    if (operand != null)
+                    {
+                        // While the optimization will technically work for any value type,
+                        // it's possible that a mutable value type has a mutating ToString,
+                        // in which case this optimization is a breaking change, due to the mutation
+                        // happening on the original value rather than on the boxed copy.
+                        // As such, we special case core types we know to be safe (and the most
+                        // important ones based on usage, anyway).
+
+                        TypeSymbol typeSymbol = operand.Type;
+                        if (typeSymbol.SpecialType.IsImmutableValueTypePrimitive() || typeSymbol.TypeKind == TypeKind.Enum)
```

Comment:
It's a good suggestion.  I was aiming to keep this fairly lean; is it cheap to look up such arbitrary types?  If so, I can do so; otherwise, I'd suggest we leave this as is for now, and it can always be updated later with more types if necessary.